[{"url": "https://api.github.com/repos/python/cpython/issues/132325", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132325/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132325/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132325/events", "html_url": "https://github.com/python/cpython/pull/132325", "id": 2983487525, "node_id": "PR_kwDOBN0Z8c6R--QT", "number": 132325, "title": "[3.13] Remove a duplicate word in the Makefile (GH-132301)", "user": {"login": "miss-islington", "id": 31488909, "node_id": "MDQ6VXNlcjMxNDg4OTA5", "avatar_url": "https://avatars.githubusercontent.com/u/31488909?v=4", "gravatar_id": "", "url": "https://api.github.com/users/miss-islington", "html_url": "https://github.com/miss-islington", "followers_url": "https://api.github.com/users/miss-islington/followers", "following_url": "https://api.github.com/users/miss-islington/following{/other_user}", "gists_url": "https://api.github.com/users/miss-islington/gists{/gist_id}", "starred_url": "https://api.github.com/users/miss-islington/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/miss-islington/subscriptions", "organizations_url": "https://api.github.com/users/miss-islington/orgs", "repos_url": "https://api.github.com/users/miss-islington/repos", "events_url": "https://api.github.com/users/miss-islington/events{/privacy}", "received_events_url": "https://api.github.com/users/miss-islington/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 647202665, "node_id": "MDU6TGFiZWw2NDcyMDI2NjU=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20issue", "name": "skip issue", "color": "000000", "default": false, "description": null}, {"id": 666979821, "node_id": "MDU6TGFiZWw2NjY5Nzk4MjE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20review", "name": "awaiting review", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T17:48:00Z", "updated_at": "2025-04-09T17:48:09Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132325", "html_url": "https://github.com/python/cpython/pull/132325", "diff_url": "https://github.com/python/cpython/pull/132325.diff", "patch_url": "https://github.com/python/cpython/pull/132325.patch", "merged_at": null}, "body": "(cherry picked from commit 4a919dd6eef1e5dbaddfd749f385f49cedc076a9)\n\nCo-authored-by: Chu <tomzz233333@gmail.com>", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132325/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132325/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132324", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132324/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132324/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132324/events", "html_url": "https://github.com/python/cpython/pull/132324", "id": 2983435482, "node_id": "PR_kwDOBN0Z8c6R-zFE", "number": 132324, "title": "gh-132307: Remove unnecessary check in `asyncio/base_events.py`", "user": {"login": "StanFromIreland", "id": 89152624, "node_id": "MDQ6VXNlcjg5MTUyNjI0", "avatar_url": "https://avatars.githubusercontent.com/u/89152624?v=4", "gravatar_id": "", "url": "https://api.github.com/users/StanFromIreland", "html_url": "https://github.com/StanFromIreland", "followers_url": "https://api.github.com/users/StanFromIreland/followers", "following_url": "https://api.github.com/users/StanFromIreland/following{/other_user}", "gists_url": "https://api.github.com/users/StanFromIreland/gists{/gist_id}", "starred_url": "https://api.github.com/users/StanFromIreland/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/StanFromIreland/subscriptions", "organizations_url": "https://api.github.com/users/StanFromIreland/orgs", "repos_url": "https://api.github.com/users/StanFromIreland/repos", "events_url": "https://api.github.com/users/StanFromIreland/events{/privacy}", "received_events_url": "https://api.github.com/users/StanFromIreland/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 666979821, "node_id": "MDU6TGFiZWw2NjY5Nzk4MjE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20review", "name": "awaiting review", "color": "fbca04", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T17:20:56Z", "updated_at": "2025-04-09T17:21:06Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132324", "html_url": "https://github.com/python/cpython/pull/132324", "diff_url": "https://github.com/python/cpython/pull/132324.diff", "patch_url": "https://github.com/python/cpython/pull/132324.patch", "merged_at": null}, "body": "<!--\r\nThanks for your contribution!\r\nPlease read this comment in its entirety. It's quite important.\r\n\r\n# Pull Request title\r\n\r\nIt should be in the following format:\r\n\r\n```\r\ngh-NNNNN: Summary of the changes made\r\n```\r\n\r\nWhere: gh-NNNNN refers to the GitHub issue number.\r\n\r\nMost PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.\r\n\r\n# Backport Pull Request title\r\n\r\nIf this is a backport PR (PR made against branches other than `main`),\r\nplease ensure that the PR title is in the following format:\r\n\r\n```\r\n[X.Y] <title from the original PR> (GH-NNNN)\r\n```\r\n\r\nWhere: [X.Y] is the branch name, e.g. [3.6].\r\n\r\nGH-NNNN refers to the PR number from `main`.\r\n\r\n-->\r\n\n\n<!-- gh-issue-number: gh-132307 -->\n* Issue: gh-132307\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132324/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132324/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132323", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132323/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132323/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132323/events", "html_url": "https://github.com/python/cpython/pull/132323", "id": 2983272002, "node_id": "PR_kwDOBN0Z8c6R-PzV", "number": 132323, "title": "gh-130645: Add colour to `argparse` help", "user": {"login": "hugovk", "id": 1324225, "node_id": "MDQ6VXNlcjEzMjQyMjU=", "avatar_url": "https://avatars.githubusercontent.com/u/1324225?v=4", "gravatar_id": "", "url": "https://api.github.com/users/hugovk", "html_url": "https://github.com/hugovk", "followers_url": "https://api.github.com/users/hugovk/followers", "following_url": "https://api.github.com/users/hugovk/following{/other_user}", "gists_url": "https://api.github.com/users/hugovk/gists{/gist_id}", "starred_url": "https://api.github.com/users/hugovk/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/hugovk/subscriptions", "organizations_url": "https://api.github.com/users/hugovk/orgs", "repos_url": "https://api.github.com/users/hugovk/repos", "events_url": "https://api.github.com/users/hugovk/events{/privacy}", "received_events_url": "https://api.github.com/users/hugovk/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T16:05:48Z", "updated_at": "2025-04-09T16:36:36Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": true, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132323", "html_url": "https://github.com/python/cpython/pull/132323", "diff_url": "https://github.com/python/cpython/pull/132323.diff", "patch_url": "https://github.com/python/cpython/pull/132323.patch", "merged_at": null}, "body": "<!--\r\nThanks for your contribution!\r\nPlease read this comment in its entirety. It's quite important.\r\n\r\n# Pull Request title\r\n\r\nIt should be in the following format:\r\n\r\n```\r\ngh-NNNNN: Summary of the changes made\r\n```\r\n\r\nWhere: gh-NNNNN refers to the GitHub issue number.\r\n\r\nMost PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.\r\n\r\n# Backport Pull Request title\r\n\r\nIf this is a backport PR (PR made against branches other than `main`),\r\nplease ensure that the PR title is in the following format:\r\n\r\n```\r\n[X.Y] <title from the original PR> (GH-NNNN)\r\n```\r\n\r\nWhere: [X.Y] is the branch name, e.g. [3.6].\r\n\r\nGH-NNNN refers to the PR number from `main`.\r\n\r\n-->\r\n\r\nAdd `color` parameter to `argparse.ArgumentParser`, which by default is `False`. When set to `True`, the help text is in colour [if allowed](https://docs.python.org/3/using/cmdline.html#using-on-controlling-color) (for example, we're not piping, `NO_COLOR=1` is not set).\r\n\r\n\r\n# Example output\r\n\r\nUsing this script:\r\n\r\n```python\r\nimport argparse\r\n\r\nparser = argparse.ArgumentParser(\r\n    description=\"calculate X to the power of Y\",\r\n    formatter_class=argparse.ArgumentDefaultsHelpFormatter,\r\n    prefix_chars=\"-+\",\r\n    prog=\"PROG\",\r\n)\r\nparser.color = True\r\ngroup = parser.add_mutually_exclusive_group()\r\ngroup.add_argument(\"-v\", \"--verbose\", action=\"store_true\", help=\"more spam\")\r\ngroup.add_argument(\"-q\", \"--quiet\", action=\"store_true\", help=\"less spam\")\r\nparser.add_argument(\"x\", type=int, help=\"the base\")\r\nparser.add_argument(\"y\", type=int, help=\"the exponent\", deprecated=True)\r\nparser.add_argument(\r\n    \"this_indeed_is_a_very_long_action_name\", type=int, help=\"the exponent\"\r\n)\r\nparser.add_argument(\"-o\", \"--optional1\", action=\"store_true\", deprecated=True)\r\nparser.add_argument(\"--optional2\", help=\"pick one\")\r\nparser.add_argument(\"--optional3\", choices=(\"X\", \"Y\", \"Z\"))\r\nparser.add_argument(\"--optional4\", choices=(\"X\", \"Y\", \"Z\"), help=\"pick one\")\r\nparser.add_argument(\"--optional5\", choices=(\"X\", \"Y\", \"Z\"), help=\"pick one\")\r\nparser.add_argument(\"--optional6\", choices=(\"X\", \"Y\", \"Z\"), help=\"pick one\")\r\nparser.add_argument(\r\n    \"-p\", \"--optional7\", choices=(\"Aaaaa\", \"Bbbbb\", \"Ccccc\", \"Ddddd\"), help=\"pick one\"\r\n)\r\n\r\nparser.add_argument(\"+f\")\r\nparser.add_argument(\"++bar\")\r\nparser.add_argument(\"-+baz\")\r\nparser.add_argument(\"-c\", \"--count\")\r\n\r\nsubparsers = parser.add_subparsers(\r\n    title=\"subcommands\", description=\"valid subcommands\", help=\"additional help\"\r\n)\r\nsubparsers.add_parser(\"sub1\", deprecated=True, help=\"sub1 help\")\r\nsub2 = subparsers.add_parser(\"sub2\", deprecated=True, help=\"sub2 help\")\r\nsub2.add_argument(\"--baz\", choices=(\"X\", \"Y\", \"Z\"), help=\"baz help\")\r\n\r\nargs = parser.parse_args()\r\n```\r\n\r\n<img width=\"864\" alt=\"image\" src=\"https://github.com/user-attachments/assets/2c210bdb-6688-4e2c-94d7-b55d7c6243ac\" />\r\n\r\n# Performance\r\n\r\nWith the same script and PGO+LTO on macOS, running it once (`time ./python.exe argparse-cli.py --help`) and 100 times (`time ./100.sh`):\r\n\r\n```sh\r\n# 100.sh\r\nfor i in $(seq 1 100); do\r\n    ./python.exe argparse-cli.py -h\r\ndone\r\n```\r\n\r\n`main` (3feac7a093b3fcd549c5dc54277f26f585f2ab0c):\r\n\r\n```\r\n./python.exe argparse-cli.py --help  0.03s user 0.01s system 88% cpu 0.046 total\r\n./100.sh  1.51s user 0.39s system 93% cpu 2.041 total\r\n```\r\n\r\nPR and colour not enabled:\r\n\r\n\r\n```\r\n./python.exe argparse-cli.py --help  0.02s user 0.02s system 58% cpu 0.072 total\r\n./100.sh  1.52s user 0.40s system 92% cpu 2.070 total\r\n```\r\n\r\nPR and colour enabled:\r\n\r\n```\r\n./python.exe argparse-cli.py --help  0.04s user 0.02s system 70% cpu 0.080 total\r\n./100.sh  1.57s user 0.40s system 93% cpu 2.106 total\r\n```\r\n\r\nSomething like 0.65 ms difference per run between before and after, averaging from `100.sh`.\n\n<!-- gh-issue-number: gh-130645 -->\n* Issue: gh-130645\n<!-- /gh-issue-number -->\n\r\n\r\n<!-- readthedocs-preview cpython-previews start -->\r\n----\n\ud83d\udcda Documentation preview \ud83d\udcda: https://cpython-previews--132323.org.readthedocs.build/\n\r\n<!-- readthedocs-preview cpython-previews end -->", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132323/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132323/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132322", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132322/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132322/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132322/events", "html_url": "https://github.com/python/cpython/issues/132322", "id": 2983190354, "node_id": "I_kwDOBN0Z8c6xz99S", "number": 132322, "title": "shutil: Inconsistent return types when using pathlib", "user": {"login": "srittau", "id": 52799, "node_id": "MDQ6VXNlcjUyNzk5", "avatar_url": "https://avatars.githubusercontent.com/u/52799?v=4", "gravatar_id": "", "url": "https://api.github.com/users/srittau", "html_url": "https://github.com/srittau", "followers_url": "https://api.github.com/users/srittau/followers", "following_url": "https://api.github.com/users/srittau/following{/other_user}", "gists_url": "https://api.github.com/users/srittau/gists{/gist_id}", "starred_url": "https://api.github.com/users/srittau/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/srittau/subscriptions", "organizations_url": "https://api.github.com/users/srittau/orgs", "repos_url": "https://api.github.com/users/srittau/repos", "events_url": "https://api.github.com/users/srittau/events{/privacy}", "received_events_url": "https://api.github.com/users/srittau/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 4255101819, "node_id": "LA_kwDOBN0Z8c79n7N7", "url": "https://api.github.com/repos/python/cpython/labels/topic-pathlib", "name": "topic-pathlib", "color": "0052cc", "default": false, "description": ""}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T15:33:00Z", "updated_at": "2025-04-09T17:30:17Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Bug report\n\n### Bug description:\n\nUsing `pathlib` with `shutil` usually works, but there are some quirks when it comes to return types. As an example, see the `copy` function (but other functions are affected as well):\n\nhttps://github.com/python/cpython/blob/67ded6a4faae29edff8e4f7886978e71ce116e33/Lib/shutil.py#L468-L484\n\nIn this case, if a `Path` object gets passed in as `dst`, it will be returned unchanged if it is not a directory, but a `str` will be returned if it is a directory. That's unexpected, the return type should be consistent, probably by always returning a `str`.\n\nI assume that pathlib support is mostly incidental in shutil, considering that the latter predates the former. shutil should probably be reviewed for proper pathlib support at some point. (See also a few other pathlib/shutil related issues reported here.)\n\n### CPython versions tested on:\n\nCPython main branch\n\n### Operating systems tested on:\n\n_No response_", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132322/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132322/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132318", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132318/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132318/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132318/events", "html_url": "https://github.com/python/cpython/pull/132318", "id": 2982877552, "node_id": "PR_kwDOBN0Z8c6R85Jd", "number": 132318, "title": "Fix a typo in c-api/unicode.rst", "user": {"login": "Yzi-Li", "id": 204532581, "node_id": "U_kgDODDDrZQ", "avatar_url": "https://avatars.githubusercontent.com/u/204532581?v=4", "gravatar_id": "", "url": "https://api.github.com/users/Yzi-Li", "html_url": "https://github.com/Yzi-Li", "followers_url": "https://api.github.com/users/Yzi-Li/followers", "following_url": "https://api.github.com/users/Yzi-Li/following{/other_user}", "gists_url": "https://api.github.com/users/Yzi-Li/gists{/gist_id}", "starred_url": "https://api.github.com/users/Yzi-Li/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/Yzi-Li/subscriptions", "organizations_url": "https://api.github.com/users/Yzi-Li/orgs", "repos_url": "https://api.github.com/users/Yzi-Li/repos", "events_url": "https://api.github.com/users/Yzi-Li/events{/privacy}", "received_events_url": "https://api.github.com/users/Yzi-Li/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}, {"id": 647202665, "node_id": "MDU6TGFiZWw2NDcyMDI2NjU=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20issue", "name": "skip issue", "color": "000000", "default": false, "description": null}, {"id": 666979871, "node_id": "MDU6TGFiZWw2NjY5Nzk4NzE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20core%20review", "name": "awaiting core review", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}, {"id": 6924885615, "node_id": "LA_kwDOBN0Z8c8AAAABnMFebw", "url": "https://api.github.com/repos/python/cpython/labels/needs%20backport%20to%203.13", "name": "needs backport to 3.13", "color": "c2e0c6", "default": false, "description": "bugs and security fixes"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T13:42:44Z", "updated_at": "2025-04-09T15:12:23Z", "closed_at": null, "author_association": "NONE", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132318", "html_url": "https://github.com/python/cpython/pull/132318", "diff_url": "https://github.com/python/cpython/pull/132318.diff", "patch_url": "https://github.com/python/cpython/pull/132318.patch", "merged_at": null}, "body": "Correct a spelling error in documentation.\r\n\r\n<!--\r\nThanks for your contribution!\r\nPlease read this comment in its entirety. It's quite important.\r\n\r\n# Pull Request title\r\n\r\nIt should be in the following format:\r\n\r\n```\r\ngh-NNNNN: Summary of the changes made\r\n```\r\n\r\nWhere: gh-NNNNN refers to the GitHub issue number.\r\n\r\nMost PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.\r\n\r\n# Backport Pull Request title\r\n\r\nIf this is a backport PR (PR made against branches other than `main`),\r\nplease ensure that the PR title is in the following format:\r\n\r\n```\r\n[X.Y] <title from the original PR> (GH-NNNN)\r\n```\r\n\r\nWhere: [X.Y] is the branch name, e.g. [3.6].\r\n\r\nGH-NNNN refers to the PR number from `main`.\r\n\r\n-->\r\n\r\n\r\n<!-- readthedocs-preview cpython-previews start -->\r\n----\n\ud83d\udcda Documentation preview \ud83d\udcda: https://cpython-previews--132318.org.readthedocs.build/\n\r\n<!-- readthedocs-preview cpython-previews end -->", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132318/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132318/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132317", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132317/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132317/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132317/events", "html_url": "https://github.com/python/cpython/pull/132317", "id": 2982871223, "node_id": "PR_kwDOBN0Z8c6R83wm", "number": 132317, "title": "Fix a typo in c-api/typeobj.rst", "user": {"login": "Yzi-Li", "id": 204532581, "node_id": "U_kgDODDDrZQ", "avatar_url": "https://avatars.githubusercontent.com/u/204532581?v=4", "gravatar_id": "", "url": "https://api.github.com/users/Yzi-Li", "html_url": "https://github.com/Yzi-Li", "followers_url": "https://api.github.com/users/Yzi-Li/followers", "following_url": "https://api.github.com/users/Yzi-Li/following{/other_user}", "gists_url": "https://api.github.com/users/Yzi-Li/gists{/gist_id}", "starred_url": "https://api.github.com/users/Yzi-Li/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/Yzi-Li/subscriptions", "organizations_url": "https://api.github.com/users/Yzi-Li/orgs", "repos_url": "https://api.github.com/users/Yzi-Li/repos", "events_url": "https://api.github.com/users/Yzi-Li/events{/privacy}", "received_events_url": "https://api.github.com/users/Yzi-Li/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}, {"id": 647202665, "node_id": "MDU6TGFiZWw2NDcyMDI2NjU=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20issue", "name": "skip issue", "color": "000000", "default": false, "description": null}, {"id": 666979871, "node_id": "MDU6TGFiZWw2NjY5Nzk4NzE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20core%20review", "name": "awaiting core review", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}, {"id": 6924885615, "node_id": "LA_kwDOBN0Z8c8AAAABnMFebw", "url": "https://api.github.com/repos/python/cpython/labels/needs%20backport%20to%203.13", "name": "needs backport to 3.13", "color": "c2e0c6", "default": false, "description": "bugs and security fixes"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T13:40:31Z", "updated_at": "2025-04-09T17:03:55Z", "closed_at": null, "author_association": "NONE", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132317", "html_url": "https://github.com/python/cpython/pull/132317", "diff_url": "https://github.com/python/cpython/pull/132317.diff", "patch_url": "https://github.com/python/cpython/pull/132317.patch", "merged_at": null}, "body": "Correct a spelling error in documentation.\r\n\r\n<!--\r\nThanks for your contribution!\r\nPlease read this comment in its entirety. It's quite important.\r\n\r\n# Pull Request title\r\n\r\nIt should be in the following format:\r\n\r\n```\r\ngh-NNNNN: Summary of the changes made\r\n```\r\n\r\nWhere: gh-NNNNN refers to the GitHub issue number.\r\n\r\nMost PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.\r\n\r\n# Backport Pull Request title\r\n\r\nIf this is a backport PR (PR made against branches other than `main`),\r\nplease ensure that the PR title is in the following format:\r\n\r\n```\r\n[X.Y] <title from the original PR> (GH-NNNN)\r\n```\r\n\r\nWhere: [X.Y] is the branch name, e.g. [3.6].\r\n\r\nGH-NNNN refers to the PR number from `main`.\r\n\r\n-->\r\n\r\n\r\n<!-- readthedocs-preview cpython-previews start -->\r\n----\n\ud83d\udcda Documentation preview \ud83d\udcda: https://cpython-previews--132317.org.readthedocs.build/\n\r\n<!-- readthedocs-preview cpython-previews end -->", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132317/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132317/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132316", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132316/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132316/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132316/events", "html_url": "https://github.com/python/cpython/issues/132316", "id": 2982643577, "node_id": "I_kwDOBN0Z8c6xx4d5", "number": 132316, "title": "Test failures when hypothesis is available but cannot reach GitHub", "user": {"login": "kulikjak", "id": 18516837, "node_id": "MDQ6VXNlcjE4NTE2ODM3", "avatar_url": "https://avatars.githubusercontent.com/u/18516837?v=4", "gravatar_id": "", "url": "https://api.github.com/users/kulikjak", "html_url": "https://github.com/kulikjak", "followers_url": "https://api.github.com/users/kulikjak/followers", "following_url": "https://api.github.com/users/kulikjak/following{/other_user}", "gists_url": "https://api.github.com/users/kulikjak/gists{/gist_id}", "starred_url": "https://api.github.com/users/kulikjak/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/kulikjak/subscriptions", "organizations_url": "https://api.github.com/users/kulikjak/orgs", "repos_url": "https://api.github.com/users/kulikjak/repos", "events_url": "https://api.github.com/users/kulikjak/events{/privacy}", "received_events_url": "https://api.github.com/users/kulikjak/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 7527065505, "node_id": "LA_kwDOBN0Z8c8AAAABwKXnoQ", "url": "https://api.github.com/repos/python/cpython/labels/infra", "name": "infra", "color": "5319e7", "default": false, "description": "CI, GitHub Actions, buildbots, Dependabot, etc."}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T12:17:48Z", "updated_at": "2025-04-09T13:45:00Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Bug report\n\n### Bug description:\n\nHi, we recently started seeing several tests (namely `test_binascii`, `test_math_property`, `test_tools` and `test_zoneinfo`) to fail due to hypothesis warnings (which were recently turned into errors with https://github.com/python/cpython/pull/131802):\n```\nhypothesis.errors.HypothesisWarning: Authorization failed when trying to download artifact from GitHub. Check that you have a valid GITHUB_TOKEN set in your environment.`\n```\nor alternatively with\n```\nhypothesis.errors.HypothesisWarning: Could not connect to GitHub to get the latest artifact.\n```\nwhen I restrict the access to the internet.\n\n\nI see that running with `\"CI\"` set in the environment fixes the issue as the problematic profile with `GitHubArtifactDatabase` is not registered, but that seems more like a lucky workaround.\n\nI am not sure about the solution - maybe it can check whether `\"GITHUB_TOKEN\"` is also available? Or maybe the warning should be ignored and not cause the tests to fail.\n\nI also think it should honor `-urlfetch` and `-network` options.\n\n\n### CPython versions tested on:\n\n3.13\n\n### Operating systems tested on:\n\nOther", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132316/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132316/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132315", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132315/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132315/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132315/events", "html_url": "https://github.com/python/cpython/issues/132315", "id": 2982580451, "node_id": "I_kwDOBN0Z8c6xxpDj", "number": 132315, "title": "_testembed.c:2312:1: warning: profile count data file not found", "user": {"login": "utkonos", "id": 2775739, "node_id": "MDQ6VXNlcjI3NzU3Mzk=", "avatar_url": "https://avatars.githubusercontent.com/u/2775739?v=4", "gravatar_id": "", "url": "https://api.github.com/users/utkonos", "html_url": "https://github.com/utkonos", "followers_url": "https://api.github.com/users/utkonos/followers", "following_url": "https://api.github.com/users/utkonos/following{/other_user}", "gists_url": "https://api.github.com/users/utkonos/gists{/gist_id}", "starred_url": "https://api.github.com/users/utkonos/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/utkonos/subscriptions", "organizations_url": "https://api.github.com/users/utkonos/orgs", "repos_url": "https://api.github.com/users/utkonos/repos", "events_url": "https://api.github.com/users/utkonos/events{/privacy}", "received_events_url": "https://api.github.com/users/utkonos/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 537044141, "node_id": "MDU6TGFiZWw1MzcwNDQxNDE=", "url": "https://api.github.com/repos/python/cpython/labels/tests", "name": "tests", "color": "09fc59", "default": false, "description": "Tests in the Lib/test dir"}, {"id": 5354314490, "node_id": "LA_kwDOBN0Z8c8AAAABPyRa-g", "url": "https://api.github.com/repos/python/cpython/labels/3.13", "name": "3.13", "color": "2e730f", "default": false, "description": "bugs and security fixes"}, {"id": 6924881042, "node_id": "LA_kwDOBN0Z8c8AAAABnMFMkg", "url": "https://api.github.com/repos/python/cpython/labels/3.14", "name": "3.14", "color": "2e730f", "default": false, "description": "new features, bugs and security fixes"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T11:53:08Z", "updated_at": "2025-04-09T14:12:58Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Bug report\n\n### Bug description:\n\n```python\ndefault: gcc -shared   -fno-semantic-interposition -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g   Modules/_blake2/blake2module.o Modules/_blake2/blake2b_impl.o Modules/_blake2/blake2s_impl.o   -o Modules/_blake2.cpython-313-x86_64-linux-gnu.so\n    default: ./Programs/_testembed.c: In function \u2018main\u2019:\n    default: ./Programs/_testembed.c:2312:1: warning: \u2018/root/Python-3.13.3/Programs/_testembed.gcda\u2019 profile count data file not found [-Wmissing-profile]\n    default:  2312 | }\n    default:       | ^\n```\n\n\n### CPython versions tested on:\n\n3.13\n\n### Operating systems tested on:\n\nLinux", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132315/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132315/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132314", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132314/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132314/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132314/events", "html_url": "https://github.com/python/cpython/issues/132314", "id": 2982575394, "node_id": "I_kwDOBN0Z8c6xxn0i", "number": 132314, "title": "pycore_call.h:166:16: warning: \u2018small_stack\u2019 may be used uninitialized", "user": {"login": "utkonos", "id": 2775739, "node_id": "MDQ6VXNlcjI3NzU3Mzk=", "avatar_url": "https://avatars.githubusercontent.com/u/2775739?v=4", "gravatar_id": "", "url": "https://api.github.com/users/utkonos", "html_url": "https://github.com/utkonos", "followers_url": "https://api.github.com/users/utkonos/followers", "following_url": "https://api.github.com/users/utkonos/following{/other_user}", "gists_url": "https://api.github.com/users/utkonos/gists{/gist_id}", "starred_url": "https://api.github.com/users/utkonos/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/utkonos/subscriptions", "organizations_url": "https://api.github.com/users/utkonos/orgs", "repos_url": "https://api.github.com/users/utkonos/repos", "events_url": "https://api.github.com/users/utkonos/events{/privacy}", "received_events_url": "https://api.github.com/users/utkonos/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 5354314490, "node_id": "LA_kwDOBN0Z8c8AAAABPyRa-g", "url": "https://api.github.com/repos/python/cpython/labels/3.13", "name": "3.13", "color": "2e730f", "default": false, "description": "bugs and security fixes"}, {"id": 6924881042, "node_id": "LA_kwDOBN0Z8c8AAAABnMFMkg", "url": "https://api.github.com/repos/python/cpython/labels/3.14", "name": "3.14", "color": "2e730f", "default": false, "description": "new features, bugs and security fixes"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 2, "created_at": "2025-04-09T11:51:06Z", "updated_at": "2025-04-09T14:12:47Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Bug report\n\n### Bug description:\n\n```python\ndefault: gcc -c -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall    -fno-semantic-interposition -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c11 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-use -fprofile-correction -I./Include/internal -I./Include/internal/mimalloc  -I. -I./Include    -DPy_BUILD_CORE -o Python/perf_jit_trampoline.o Python/perf_jit_trampoline.c\n    default: In file included from Python/bltinmodule.c:5:\n    default: In function \u2018_PyObject_VectorcallTstate\u2019,\n    default:     inlined from \u2018_PyObject_VectorcallTstate\u2019 at ./Include/internal/pycore_call.h:153:1,\n    default:     inlined from \u2018map_next\u2019 at Python/bltinmodule.c:1440:14:\n    default: ./Include/internal/pycore_call.h:166:16: warning: \u2018small_stack\u2019 may be used uninitialized [-Wmaybe-uninitialized]\n    default:   166 |         return _PyObject_MakeTpCall(tstate, callable, args, nargs, kwnames);\n    default:       |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n    default: ./Include/internal/pycore_call.h: In function \u2018map_next\u2019:\n    default: ./Include/internal/pycore_call.h:107:23: note: by argument 3 of type \u2018struct PyObject * const *\u2019 to \u2018_PyObject_MakeTpCall\u2019 declared here\n    default:   107 | PyAPI_FUNC(PyObject*) _PyObject_MakeTpCall(\n    default:       |                       ^~~~~~~~~~~~~~~~~~~~\n    default: Python/bltinmodule.c:1412:15: note: \u2018small_stack\u2019 declared here\n    default:  1412 |     PyObject *small_stack[_PY_FASTCALL_SMALL_STACK];\n    default:       |               ^~~~~~~~~~~\n\n```\n\n\n### CPython versions tested on:\n\n3.13\n\n### Operating systems tested on:\n\nLinux", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132314/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132314/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132312", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132312/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132312/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132312/events", "html_url": "https://github.com/python/cpython/issues/132312", "id": 2982569939, "node_id": "I_kwDOBN0Z8c6xxmfT", "number": 132312, "title": "Add a `PyThreadState *` parameter (almost) everywhere", "user": {"login": "markshannon", "id": 9448417, "node_id": "MDQ6VXNlcjk0NDg0MTc=", "avatar_url": "https://avatars.githubusercontent.com/u/9448417?v=4", "gravatar_id": "", "url": "https://api.github.com/users/markshannon", "html_url": "https://github.com/markshannon", "followers_url": "https://api.github.com/users/markshannon/followers", "following_url": "https://api.github.com/users/markshannon/following{/other_user}", "gists_url": "https://api.github.com/users/markshannon/gists{/gist_id}", "starred_url": "https://api.github.com/users/markshannon/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/markshannon/subscriptions", "organizations_url": "https://api.github.com/users/markshannon/orgs", "repos_url": "https://api.github.com/users/markshannon/repos", "events_url": "https://api.github.com/users/markshannon/events{/privacy}", "received_events_url": "https://api.github.com/users/markshannon/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871772, "node_id": "MDU6TGFiZWw1MzY4NzE3NzI=", "url": "https://api.github.com/repos/python/cpython/labels/type-feature", "name": "type-feature", "color": "006b75", "default": false, "description": "A feature request or enhancement"}, {"id": 537029789, "node_id": "MDU6TGFiZWw1MzcwMjk3ODk=", "url": "https://api.github.com/repos/python/cpython/labels/performance", "name": "performance", "color": "0052cc", "default": false, "description": "Performance or resource usage"}, {"id": 4019060898, "node_id": "LA_kwDOBN0Z8c7vjgCi", "url": "https://api.github.com/repos/python/cpython/labels/topic-C-API", "name": "topic-C-API", "color": "0052cc", "default": false, "description": ""}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T11:49:36Z", "updated_at": "2025-04-09T17:09:39Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "This is part feature request, part performance issue and part a general appeal for assistance.\n\nWe should add a new variant taking a `PyThreadState *` parameter for most C API and internal functions.\n\nThere are two motivations for this, performance and a future, consistent C API.\n\n### Performance\n\nThe `PyThreadState` struct is ubiquitous in the VM, it controls stack usage, holds the freelists, holds the current exception, etc, etc.\n\nConsequently many C functions, both API and internal, take a `PyThreadState *tstate` parameter.\nHowever, for historical reasons, many C functions, both API and internal, do not such a parameter.\n\nThis leads to some fairly easy to fix inefficiencies, where `spam()` and `eggs()` take a thread state, but `ham()` does not, then `spam()` calls `ham()` which calls `eggs()`, forcing `ham()` to load the thread state from thread local storage, in order to pass it to `eggs()`.\nAdding a `PyThreadState *tstate` to `ham()` avoids need to access thread local storage.\n\n### Consistent, portable, future looking C API\n\nIn order to support things like tagged integers, we are going to need to a new C API.\n\nIt is out of scope to discuss what such an API would look like, but all realistic proposal so far take a \"context\" parameter to all, or almost all, API functions. Using a `PyThreadState *` parameter everywhere would make a mechanical transformation from old to new API much simpler.\n\nUnfortunately the C API is large and cannot be changed, so we will need many new functions, like `ham_tstate()` which replicates `ham()` but with a thread state parameter.\n\nThis work is largely mechanical, and can be done by inexperienced contributors. Hence the appeal for assistance.\n\n\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132312/reactions", "total_count": 1, "+1": 1, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132312/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132308", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132308/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132308/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132308/events", "html_url": "https://github.com/python/cpython/issues/132308", "id": 2982308932, "node_id": "I_kwDOBN0Z8c6xwmxE", "number": 132308, "title": "`TracebackException` swallows `__cause__` and `__context__` attributes of falsey exceptions.", "user": {"login": "YvesDup", "id": 6217656, "node_id": "MDQ6VXNlcjYyMTc2NTY=", "avatar_url": "https://avatars.githubusercontent.com/u/6217656?v=4", "gravatar_id": "", "url": "https://api.github.com/users/YvesDup", "html_url": "https://github.com/YvesDup", "followers_url": "https://api.github.com/users/YvesDup/followers", "following_url": "https://api.github.com/users/YvesDup/following{/other_user}", "gists_url": "https://api.github.com/users/YvesDup/gists{/gist_id}", "starred_url": "https://api.github.com/users/YvesDup/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/YvesDup/subscriptions", "organizations_url": "https://api.github.com/users/YvesDup/orgs", "repos_url": "https://api.github.com/users/YvesDup/repos", "events_url": "https://api.github.com/users/YvesDup/events{/privacy}", "received_events_url": "https://api.github.com/users/YvesDup/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 4018668725, "node_id": "LA_kwDOBN0Z8c7viAS1", "url": "https://api.github.com/repos/python/cpython/labels/stdlib", "name": "stdlib", "color": "09fc59", "default": false, "description": "Python modules in the Lib dir"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T10:06:24Z", "updated_at": "2025-04-09T14:13:37Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Bug report\n\n### Bug description:\n\n# Description\n\n```\nimport traceback as tb\n\nclass FalseyLenException(Exception):\n    def __len__(self):\n        return 0\n\nclass FalseyBoolException(Exception):\n    def __bool__(self):\n        return False\n\ndef tracebackexception_swallows_cause_and_context():\n    try:\n        raise FalseyBoolException(\"Oh\") from KeyError\n    except Exception as e:\n        print(f'{e = }, {e.__cause__ = }')\n        tb.print_exception(e)\n    print(\"---\"*20)\n\n    try:\n        try:\n            1/0\n        except:\n            raise FalseyLenException(\"Ah\")\n    except Exception as e:\n        print(f'{e = }, {e.__context__ = }')\n        tb.print_exception(e)\n    print(\"---\"*20)\n\nif __name__ == \"__main__\":\n    tracebackexception_swallows_cause_and_context()\n```\nOuptput is:\n```\ne = FalseyBoolException('Oh'), e.__cause__ = KeyError()\nTraceback (most recent call last):\n  File \"/Users/yves/Desktop/Bugs/gh-132XXX.py\", line 13, in tracebackexception_swallows_cause_and_context\n    raise FalseyBoolException(\"Oh\") from KeyError\nFalseyBoolException: Oh\n------------------------------------------------------------\ne = FalseyLenException('Ah'), e.__context__ = ZeroDivisionError('division by zero')\nTraceback (most recent call last):\n  File \"/Users/yves/Desktop/Bugs/gh-132XXX.py\", line 23, in tracebackexception_swallows_cause_and_context\n    raise FalseyLenException(\"Ah\")\nFalseyLenException: Ah\n```\n\nThe **Cause** or **Context** descriptions are not printed as expected.\n\n# Where Does it from ?\n\nThese following tests are wrong. \n\nhttps://github.com/python/cpython/blob/255eb375a7cc3cb16698bebf42689e8bc783ed93/Lib/traceback.py#L1123\nIt should be: `if (e is not None and e.__cause__ is not None` \nhttps://github.com/python/cpython/blob/255eb375a7cc3cb16698bebf42689e8bc783ed93/Lib/traceback.py#L1144\nIt should be: `if (e is not None and  and e.__context__ is not None` \n\n# Linked Issues\nSee: https://github.com/python/cpython/pull/132129\n\n\n### CPython versions tested on:\n\nCPython main branch\n\n### Operating systems tested on:\n\nmacOS", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132308/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132308/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132307", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132307/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132307/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132307/events", "html_url": "https://github.com/python/cpython/issues/132307", "id": 2982191433, "node_id": "I_kwDOBN0Z8c6xwKFJ", "number": 132307, "title": "Inconsistent code found by static analyzers at base_events.py", "user": {"login": "Anchels", "id": 42744001, "node_id": "MDQ6VXNlcjQyNzQ0MDAx", "avatar_url": "https://avatars.githubusercontent.com/u/42744001?v=4", "gravatar_id": "", "url": "https://api.github.com/users/Anchels", "html_url": "https://github.com/Anchels", "followers_url": "https://api.github.com/users/Anchels/followers", "following_url": "https://api.github.com/users/Anchels/following{/other_user}", "gists_url": "https://api.github.com/users/Anchels/gists{/gist_id}", "starred_url": "https://api.github.com/users/Anchels/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/Anchels/subscriptions", "organizations_url": "https://api.github.com/users/Anchels/orgs", "repos_url": "https://api.github.com/users/Anchels/repos", "events_url": "https://api.github.com/users/Anchels/events{/privacy}", "received_events_url": "https://api.github.com/users/Anchels/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 557226369, "node_id": "MDU6TGFiZWw1NTcyMjYzNjk=", "url": "https://api.github.com/repos/python/cpython/labels/topic-asyncio", "name": "topic-asyncio", "color": "0052cc", "default": false, "description": ""}, {"id": 4018668725, "node_id": "LA_kwDOBN0Z8c7viAS1", "url": "https://api.github.com/repos/python/cpython/labels/stdlib", "name": "stdlib", "color": "09fc59", "default": false, "description": "Python modules in the Lib dir"}, {"id": 4018743106, "node_id": "LA_kwDOBN0Z8c7viSdC", "url": "https://api.github.com/repos/python/cpython/labels/easy", "name": "easy", "color": "0e8a16", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T09:31:36Z", "updated_at": "2025-04-09T17:21:10Z", "closed_at": null, "author_association": "NONE", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "Greetings! I've been analyzing Cpython with Svace static analyzer. It has found a code inconsitency issue at the `asyncio` library in the following method:\n\nhttps://github.com/python/cpython/blob/7ebbd271444d89218870169624921b795a717470/Lib/asyncio/base_events.py#L1653-L1681\n\nThe problem is a redundant comparison with a `None` value for reference `sock` here:\nhttps://github.com/python/cpython/blob/7ebbd271444d89218870169624921b795a717470/Lib/asyncio/base_events.py#L1669-L1670\n\nwhich has already been dereferenced before:\nhttps://github.com/python/cpython/blob/7ebbd271444d89218870169624921b795a717470/Lib/asyncio/base_events.py#L1658-L1659\n\nAccording to the [documentation](https://docs.python.org/3/library/asyncio-eventloop.html#:~:text=sock%20is%20a%20preexisting%20socket%20object%20returned%20from%20socket.accept), the `sock` input parameter is a preexisting object returned from the `socket.accept()` function. So it shouldn't be `None`\n\n---\n### Proposed solution\n\nI believe it is advisable to either move the null check for `sock` to the beginning of the function, or remove it. I'm not sure which of these options would be more in line with the regulations. After I get a response, I might be able to create a pull request.\n\n<!-- gh-linked-prs -->\n### Linked PRs\n* gh-132324\n<!-- /gh-linked-prs -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132307/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132307/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132306", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132306/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132306/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132306/events", "html_url": "https://github.com/python/cpython/pull/132306", "id": 2982128152, "node_id": "PR_kwDOBN0Z8c6R6Uwc", "number": 132306, "title": "gh-132305: Make Argument Clinic code compatible with Python 3.10", "user": {"login": "serhiy-storchaka", "id": 3659035, "node_id": "MDQ6VXNlcjM2NTkwMzU=", "avatar_url": "https://avatars.githubusercontent.com/u/3659035?v=4", "gravatar_id": "", "url": "https://api.github.com/users/serhiy-storchaka", "html_url": "https://github.com/serhiy-storchaka", "followers_url": "https://api.github.com/users/serhiy-storchaka/followers", "following_url": "https://api.github.com/users/serhiy-storchaka/following{/other_user}", "gists_url": "https://api.github.com/users/serhiy-storchaka/gists{/gist_id}", "starred_url": "https://api.github.com/users/serhiy-storchaka/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/serhiy-storchaka/subscriptions", "organizations_url": "https://api.github.com/users/serhiy-storchaka/orgs", "repos_url": "https://api.github.com/users/serhiy-storchaka/repos", "events_url": "https://api.github.com/users/serhiy-storchaka/events{/privacy}", "received_events_url": "https://api.github.com/users/serhiy-storchaka/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 666980059, "node_id": "MDU6TGFiZWw2NjY5ODAwNTk=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20merge", "name": "awaiting merge", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T09:12:58Z", "updated_at": "2025-04-09T13:40:39Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132306", "html_url": "https://github.com/python/cpython/pull/132306", "diff_url": "https://github.com/python/cpython/pull/132306.diff", "patch_url": "https://github.com/python/cpython/pull/132306.patch", "merged_at": null}, "body": "\n\n<!-- gh-issue-number: gh-132305 -->\n* Issue: gh-132305\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132306/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132306/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132305", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132305/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132305/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132305/events", "html_url": "https://github.com/python/cpython/issues/132305", "id": 2982113821, "node_id": "I_kwDOBN0Z8c6xv3Id", "number": 132305, "title": "Argument Clinic code is incompatible with older Python", "user": {"login": "serhiy-storchaka", "id": 3659035, "node_id": "MDQ6VXNlcjM2NTkwMzU=", "avatar_url": "https://avatars.githubusercontent.com/u/3659035?v=4", "gravatar_id": "", "url": "https://api.github.com/users/serhiy-storchaka", "html_url": "https://github.com/serhiy-storchaka", "followers_url": "https://api.github.com/users/serhiy-storchaka/followers", "following_url": "https://api.github.com/users/serhiy-storchaka/following{/other_user}", "gists_url": "https://api.github.com/users/serhiy-storchaka/gists{/gist_id}", "starred_url": "https://api.github.com/users/serhiy-storchaka/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/serhiy-storchaka/subscriptions", "organizations_url": "https://api.github.com/users/serhiy-storchaka/orgs", "repos_url": "https://api.github.com/users/serhiy-storchaka/repos", "events_url": "https://api.github.com/users/serhiy-storchaka/events{/privacy}", "received_events_url": "https://api.github.com/users/serhiy-storchaka/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 4018676104, "node_id": "LA_kwDOBN0Z8c7viCGI", "url": "https://api.github.com/repos/python/cpython/labels/build", "name": "build", "color": "0052cc", "default": false, "description": "The build process and cross-build"}, {"id": 4019554242, "node_id": "LA_kwDOBN0Z8c7vlYfC", "url": "https://api.github.com/repos/python/cpython/labels/topic-argument-clinic", "name": "topic-argument-clinic", "color": "0052cc", "default": false, "description": ""}, {"id": 6924881042, "node_id": "LA_kwDOBN0Z8c8AAAABnMFMkg", "url": "https://api.github.com/repos/python/cpython/labels/3.14", "name": "3.14", "color": "2e730f", "default": false, "description": "new features, bugs and security fixes"}], "state": "open", "locked": false, "assignee": {"login": "serhiy-storchaka", "id": 3659035, "node_id": "MDQ6VXNlcjM2NTkwMzU=", "avatar_url": "https://avatars.githubusercontent.com/u/3659035?v=4", "gravatar_id": "", "url": "https://api.github.com/users/serhiy-storchaka", "html_url": "https://github.com/serhiy-storchaka", "followers_url": "https://api.github.com/users/serhiy-storchaka/followers", "following_url": "https://api.github.com/users/serhiy-storchaka/following{/other_user}", "gists_url": "https://api.github.com/users/serhiy-storchaka/gists{/gist_id}", "starred_url": "https://api.github.com/users/serhiy-storchaka/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/serhiy-storchaka/subscriptions", "organizations_url": "https://api.github.com/users/serhiy-storchaka/orgs", "repos_url": "https://api.github.com/users/serhiy-storchaka/repos", "events_url": "https://api.github.com/users/serhiy-storchaka/events{/privacy}", "received_events_url": "https://api.github.com/users/serhiy-storchaka/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "assignees": [{"login": "serhiy-storchaka", "id": 3659035, "node_id": "MDQ6VXNlcjM2NTkwMzU=", "avatar_url": "https://avatars.githubusercontent.com/u/3659035?v=4", "gravatar_id": "", "url": "https://api.github.com/users/serhiy-storchaka", "html_url": "https://github.com/serhiy-storchaka", "followers_url": "https://api.github.com/users/serhiy-storchaka/followers", "following_url": "https://api.github.com/users/serhiy-storchaka/following{/other_user}", "gists_url": "https://api.github.com/users/serhiy-storchaka/gists{/gist_id}", "starred_url": "https://api.github.com/users/serhiy-storchaka/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/serhiy-storchaka/subscriptions", "organizations_url": "https://api.github.com/users/serhiy-storchaka/orgs", "repos_url": "https://api.github.com/users/serhiy-storchaka/repos", "events_url": "https://api.github.com/users/serhiy-storchaka/events{/privacy}", "received_events_url": "https://api.github.com/users/serhiy-storchaka/received_events", "type": "User", "user_view_type": "public", "site_admin": false}], "milestone": null, "comments": 1, "created_at": "2025-04-09T09:09:14Z", "updated_at": "2025-04-09T09:20:37Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "Until recently I used Python 3.11 to regenerate Argument Clinic code. But `make clinic` stopped to work  because Python code for some Argument Clinic converted uses syntax incompatible with 3.11. We should be conservative with Python code that is used to build Python.\n\n<!-- gh-linked-prs -->\n### Linked PRs\n* gh-132306\n<!-- /gh-linked-prs -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132305/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132305/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132304", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132304/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132304/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132304/events", "html_url": "https://github.com/python/cpython/pull/132304", "id": 2982069361, "node_id": "PR_kwDOBN0Z8c6R6Iz5", "number": 132304, "title": "gh-130587: Invoke regen-token rst with rstfile as an argument", "user": {"login": "befeleme", "id": 33810531, "node_id": "MDQ6VXNlcjMzODEwNTMx", "avatar_url": "https://avatars.githubusercontent.com/u/33810531?v=4", "gravatar_id": "", "url": "https://api.github.com/users/befeleme", "html_url": "https://github.com/befeleme", "followers_url": "https://api.github.com/users/befeleme/followers", "following_url": "https://api.github.com/users/befeleme/following{/other_user}", "gists_url": "https://api.github.com/users/befeleme/gists{/gist_id}", "starred_url": "https://api.github.com/users/befeleme/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/befeleme/subscriptions", "organizations_url": "https://api.github.com/users/befeleme/orgs", "repos_url": "https://api.github.com/users/befeleme/repos", "events_url": "https://api.github.com/users/befeleme/events{/privacy}", "received_events_url": "https://api.github.com/users/befeleme/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 666980059, "node_id": "MDU6TGFiZWw2NjY5ODAwNTk=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20merge", "name": "awaiting merge", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}, {"id": 6924885615, "node_id": "LA_kwDOBN0Z8c8AAAABnMFebw", "url": "https://api.github.com/repos/python/cpython/labels/needs%20backport%20to%203.13", "name": "needs backport to 3.13", "color": "c2e0c6", "default": false, "description": "bugs and security fixes"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T08:53:30Z", "updated_at": "2025-04-09T17:00:47Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132304", "html_url": "https://github.com/python/cpython/pull/132304", "diff_url": "https://github.com/python/cpython/pull/132304.diff", "patch_url": "https://github.com/python/cpython/pull/132304.patch", "merged_at": null}, "body": "This mitigates the issue we've bumped into in Fedora where we create special build directories for various types of builds and we can't deal with relative paths efficiently. I believe this was omitted incidentally in the original change.\r\nSpotted in Python 3.14.0a7.\n\n<!-- gh-issue-number: gh-130587 -->\n* Issue: gh-130587\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132304/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132304/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132300", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132300/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132300/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132300/events", "html_url": "https://github.com/python/cpython/pull/132300", "id": 2981565980, "node_id": "PR_kwDOBN0Z8c6R4cIo", "number": 132300, "title": "doc: true division doesn't use co_flags anymore", "user": {"login": "methane", "id": 199592, "node_id": "MDQ6VXNlcjE5OTU5Mg==", "avatar_url": "https://avatars.githubusercontent.com/u/199592?v=4", "gravatar_id": "", "url": "https://api.github.com/users/methane", "html_url": "https://github.com/methane", "followers_url": "https://api.github.com/users/methane/followers", "following_url": "https://api.github.com/users/methane/following{/other_user}", "gists_url": "https://api.github.com/users/methane/gists{/gist_id}", "starred_url": "https://api.github.com/users/methane/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/methane/subscriptions", "organizations_url": "https://api.github.com/users/methane/orgs", "repos_url": "https://api.github.com/users/methane/repos", "events_url": "https://api.github.com/users/methane/events{/privacy}", "received_events_url": "https://api.github.com/users/methane/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 536871770, "node_id": "MDU6TGFiZWw1MzY4NzE3NzA=", "url": "https://api.github.com/repos/python/cpython/labels/type-bug", "name": "type-bug", "color": "e99695", "default": false, "description": "An unexpected behavior, bug, or error"}, {"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}, {"id": 647202665, "node_id": "MDU6TGFiZWw2NDcyMDI2NjU=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20issue", "name": "skip issue", "color": "000000", "default": false, "description": null}, {"id": 666979871, "node_id": "MDU6TGFiZWw2NjY5Nzk4NzE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20core%20review", "name": "awaiting core review", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}, {"id": 6924885615, "node_id": "LA_kwDOBN0Z8c8AAAABnMFebw", "url": "https://api.github.com/repos/python/cpython/labels/needs%20backport%20to%203.13", "name": "needs backport to 3.13", "color": "c2e0c6", "default": false, "description": "bugs and security fixes"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-09T05:21:24Z", "updated_at": "2025-04-09T05:21:40Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132300", "html_url": "https://github.com/python/cpython/pull/132300", "diff_url": "https://github.com/python/cpython/pull/132300.diff", "patch_url": "https://github.com/python/cpython/pull/132300.patch", "merged_at": null}, "body": "\r\n\r\n<!-- readthedocs-preview cpython-previews start -->\r\n----\n\ud83d\udcda Documentation preview \ud83d\udcda: https://cpython-previews--132300.org.readthedocs.build/\n\r\n<!-- readthedocs-preview cpython-previews end -->", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132300/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132300/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132299", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132299/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132299/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132299/events", "html_url": "https://github.com/python/cpython/issues/132299", "id": 2981559938, "node_id": "I_kwDOBN0Z8c6xtv6C", "number": 132299, "title": "Call built-in method of object instance, lost self param?", "user": {"login": "A4-Tacks", "id": 102709083, "node_id": "U_kgDOBh83Ww", "avatar_url": "https://avatars.githubusercontent.com/u/102709083?v=4", "gravatar_id": "", "url": "https://api.github.com/users/A4-Tacks", "html_url": "https://github.com/A4-Tacks", "followers_url": "https://api.github.com/users/A4-Tacks/followers", "following_url": "https://api.github.com/users/A4-Tacks/following{/other_user}", "gists_url": "https://api.github.com/users/A4-Tacks/gists{/gist_id}", "starred_url": "https://api.github.com/users/A4-Tacks/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/A4-Tacks/subscriptions", "organizations_url": "https://api.github.com/users/A4-Tacks/orgs", "repos_url": "https://api.github.com/users/A4-Tacks/repos", "events_url": "https://api.github.com/users/A4-Tacks/events{/privacy}", "received_events_url": "https://api.github.com/users/A4-Tacks/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}], "state": "open", "locked": false, "assignee": {"login": "skirpichev", "id": 2155800, "node_id": "MDQ6VXNlcjIxNTU4MDA=", "avatar_url": "https://avatars.githubusercontent.com/u/2155800?v=4", "gravatar_id": "", "url": "https://api.github.com/users/skirpichev", "html_url": "https://github.com/skirpichev", "followers_url": "https://api.github.com/users/skirpichev/followers", "following_url": "https://api.github.com/users/skirpichev/following{/other_user}", "gists_url": "https://api.github.com/users/skirpichev/gists{/gist_id}", "starred_url": "https://api.github.com/users/skirpichev/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/skirpichev/subscriptions", "organizations_url": "https://api.github.com/users/skirpichev/orgs", "repos_url": "https://api.github.com/users/skirpichev/repos", "events_url": "https://api.github.com/users/skirpichev/events{/privacy}", "received_events_url": "https://api.github.com/users/skirpichev/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "assignees": [{"login": "skirpichev", "id": 2155800, "node_id": "MDQ6VXNlcjIxNTU4MDA=", "avatar_url": "https://avatars.githubusercontent.com/u/2155800?v=4", "gravatar_id": "", "url": "https://api.github.com/users/skirpichev", "html_url": "https://github.com/skirpichev", "followers_url": "https://api.github.com/users/skirpichev/followers", "following_url": "https://api.github.com/users/skirpichev/following{/other_user}", "gists_url": "https://api.github.com/users/skirpichev/gists{/gist_id}", "starred_url": "https://api.github.com/users/skirpichev/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/skirpichev/subscriptions", "organizations_url": "https://api.github.com/users/skirpichev/orgs", "repos_url": "https://api.github.com/users/skirpichev/repos", "events_url": "https://api.github.com/users/skirpichev/events{/privacy}", "received_events_url": "https://api.github.com/users/skirpichev/received_events", "type": "User", "user_view_type": "public", "site_admin": false}], "milestone": null, "comments": 1, "created_at": "2025-04-09T05:16:52Z", "updated_at": "2025-04-09T06:51:18Z", "closed_at": null, "author_association": "NONE", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "```python\n>>> class Foo:\n...  f=print\n...\n>>> Foo.f(2)\n2\n>>> Foo().f(2)\n2\n>>> def f(*args): print(*args)\n...\n>>> class Foo:\n...  f=f\n...\n>>> Foo.f(2)\n2\n>>> Foo().f(2)\n<__main__.Foo object at 0x7b16f5e810> 2\n```\n\nIs there any difference between f and print? Is this a bug?\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132299/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132299/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132298", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132298/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132298/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132298/events", "html_url": "https://github.com/python/cpython/pull/132298", "id": 2981357538, "node_id": "PR_kwDOBN0Z8c6R3v-_", "number": 132298, "title": "gh-132296: Use thread-state reference counting to fix `PyEval_SetTraceAllThreads`", "user": {"login": "ZeroIntensity", "id": 49501366, "node_id": "MDQ6VXNlcjQ5NTAxMzY2", "avatar_url": "https://avatars.githubusercontent.com/u/49501366?v=4", "gravatar_id": "", "url": "https://api.github.com/users/ZeroIntensity", "html_url": "https://github.com/ZeroIntensity", "followers_url": "https://api.github.com/users/ZeroIntensity/followers", "following_url": "https://api.github.com/users/ZeroIntensity/following{/other_user}", "gists_url": "https://api.github.com/users/ZeroIntensity/gists{/gist_id}", "starred_url": "https://api.github.com/users/ZeroIntensity/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/ZeroIntensity/subscriptions", "organizations_url": "https://api.github.com/users/ZeroIntensity/orgs", "repos_url": "https://api.github.com/users/ZeroIntensity/repos", "events_url": "https://api.github.com/users/ZeroIntensity/events{/privacy}", "received_events_url": "https://api.github.com/users/ZeroIntensity/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 6140426713, "node_id": "LA_kwDOBN0Z8c8AAAABbf952Q", "url": "https://api.github.com/repos/python/cpython/labels/topic-free-threading", "name": "topic-free-threading", "color": "0052cc", "default": false, "description": ""}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T02:17:00Z", "updated_at": "2025-04-09T02:27:29Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": true, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132298", "html_url": "https://github.com/python/cpython/pull/132298", "diff_url": "https://github.com/python/cpython/pull/132298.diff", "patch_url": "https://github.com/python/cpython/pull/132298.patch", "merged_at": null}, "body": "This extends the `refcount` field to affect `PyThreadState_Delete` and friends, letting us keep a thread state alive for slightly longer. This mechanism is far from being robust, because we're still not preventing the thread state from being cleared or other bad things, but it's a really nice start.\r\n\r\nWe *could* go the route of letting the reference count field prevent any finalization happening to the thread state, but I want to hear opinions on this approach first.\n\n<!-- gh-issue-number: gh-132296 -->\n* Issue: gh-132296\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132298/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132298/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132296", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132296/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132296/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132296/events", "html_url": "https://github.com/python/cpython/issues/132296", "id": 2981203176, "node_id": "I_kwDOBN0Z8c6xsYzo", "number": 132296, "title": "Concurrent deallocation of threads while calling `PyEval_SetTrace`", "user": {"login": "devdanzin", "id": 74280297, "node_id": "MDQ6VXNlcjc0MjgwMjk3", "avatar_url": "https://avatars.githubusercontent.com/u/74280297?v=4", "gravatar_id": "", "url": "https://api.github.com/users/devdanzin", "html_url": "https://github.com/devdanzin", "followers_url": "https://api.github.com/users/devdanzin/followers", "following_url": "https://api.github.com/users/devdanzin/following{/other_user}", "gists_url": "https://api.github.com/users/devdanzin/gists{/gist_id}", "starred_url": "https://api.github.com/users/devdanzin/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/devdanzin/subscriptions", "organizations_url": "https://api.github.com/users/devdanzin/orgs", "repos_url": "https://api.github.com/users/devdanzin/repos", "events_url": "https://api.github.com/users/devdanzin/events{/privacy}", "received_events_url": "https://api.github.com/users/devdanzin/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 4018668718, "node_id": "LA_kwDOBN0Z8c7viASu", "url": "https://api.github.com/repos/python/cpython/labels/interpreter-core", "name": "interpreter-core", "color": "09fc59", "default": false, "description": "(Objects, Python, Grammar, and Parser dirs)"}, {"id": 4018848914, "node_id": "LA_kwDOBN0Z8c7visSS", "url": "https://api.github.com/repos/python/cpython/labels/type-crash", "name": "type-crash", "color": "b60205", "default": false, "description": "A hard crash of the interpreter, possibly with a core dump"}, {"id": 4019060898, "node_id": "LA_kwDOBN0Z8c7vjgCi", "url": "https://api.github.com/repos/python/cpython/labels/topic-C-API", "name": "topic-C-API", "color": "0052cc", "default": false, "description": ""}, {"id": 6140426713, "node_id": "LA_kwDOBN0Z8c8AAAABbf952Q", "url": "https://api.github.com/repos/python/cpython/labels/topic-free-threading", "name": "topic-free-threading", "color": "0052cc", "default": false, "description": ""}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-09T00:23:50Z", "updated_at": "2025-04-09T12:25:21Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Crash report\n\n### What happened?\n\nThe following code will cause the interpreter to abort in a no-gil build:\n```python\nimport threading\n\nfor x in range(1000):\n    threading._start_joinable_thread(lambda: None)\n    try:\n        threading.settrace_all_threads(())\n    except Exception:\n        pass\n```\n\nAbort backtrace:\n```gdb\npython: Python/legacy_tracing.c:435: is_tstate_valid: Assertion `!_PyMem_IsPtrFreed(tstate)' failed.\nThread 1 \"python\" received signal SIGABRT, Aborted.\n\n__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737350580032) at ./nptl/pthread_kill.c:44\n44      ./nptl/pthread_kill.c: No such file or directory.\n\n#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737350580032) at ./nptl/pthread_kill.c:44\n#1  __pthread_kill_internal (signo=6, threadid=140737350580032) at ./nptl/pthread_kill.c:78\n#2  __GI___pthread_kill (threadid=140737350580032, signo=signo@entry=6) at ./nptl/pthread_kill.c:89\n#3  0x00007ffff7ce0476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26\n#4  0x00007ffff7cc67f3 in __GI_abort () at ./stdlib/abort.c:79\n#5  0x00007ffff7cc671b in __assert_fail_base (fmt=0x7ffff7e7b130 \"%s%s%s:%u: %s%sAssertion `%s' failed.\\n%n\",\n    assertion=0x555555b53aa9 \"!_PyMem_IsPtrFreed(tstate)\", file=0x555555b53a91 \"Python/legacy_tracing.c\", line=435,\n    function=<optimized out>) at ./assert/assert.c:94\n#6  0x00007ffff7cd7e96 in __GI___assert_fail (assertion=assertion@entry=0x555555b53aa9 \"!_PyMem_IsPtrFreed(tstate)\",\n    file=file@entry=0x555555b53a91 \"Python/legacy_tracing.c\", line=line@entry=435,\n    function=function@entry=0x555555b53ed0 <__PRETTY_FUNCTION__.18> \"is_tstate_valid\") at ./assert/assert.c:103\n#7  0x000055555594a667 in is_tstate_valid (tstate=tstate@entry=0xdddddddddddddddd) at Python/legacy_tracing.c:435\n#8  0x000055555594c0c2 in _PyEval_SetTrace (tstate=tstate@entry=0xdddddddddddddddd,\n    func=func@entry=0x5555559984b0 <trace_trampoline>, arg=arg@entry=()) at Python/legacy_tracing.c:596\n#9  0x000055555584cd32 in PyEval_SetTraceAllThreads (func=0x5555559984b0 <trace_trampoline>, arg=()) at Python/ceval.c:2473\n#10 0x000055555599698e in sys__settraceallthreads (module=<optimized out>, arg=<optimized out>) at ./Python/sysmodule.c:1187\n#11 0x0000555555859272 in _PyEval_EvalFrameDefault (tstate=tstate@entry=0x555555d8a858 <_PyRuntime+361432>, frame=0x7ffff7fb0098,\n    frame@entry=0x7ffff7fb0020, throwflag=throwflag@entry=0) at Python/generated_cases.c.h:2271\n#12 0x0000555555888b6d in _PyEval_EvalFrame (throwflag=0, frame=0x7ffff7fb0020, tstate=0x555555d8a858 <_PyRuntime+361432>)\n    at ./Include/internal/pycore_ceval.h:119\n```\n\nIt's also possible to sometimes segfault the interpreter with similar code, like:\n```python\nimport threading\n\nthreading._start_joinable_thread(lambda: None)\nthreading._start_joinable_thread(lambda: None)\nthreading._start_joinable_thread(lambda: None)\nthreading._start_joinable_thread(lambda: None)\nfor x in range(100):\n    try:\n        threading.settrace_all_threads(())\n    except Exception:\n        pass\n```\n\nSegfault backtrace:\n```gdb\nThread 1 \"python\" received signal SIGSEGV, Segmentation fault.\n0x000055555594b3fb in setup_tracing (tstate=tstate@entry=0x555555e29760, func=func@entry=0x5555559984b0 <trace_trampoline>, arg=arg@entry=(), old_traceobj=old_traceobj@entry=0x7fffffffd380) at Python/legacy_tracing.c:588\n588         tstate->interp->sys_tracing_threads += delta;\n(gdb) bt\n#0  0x000055555594b3fb in setup_tracing (tstate=tstate@entry=0x555555e29760, func=func@entry=0x5555559984b0 <trace_trampoline>,\n    arg=arg@entry=(), old_traceobj=old_traceobj@entry=0x7fffffffd380) at Python/legacy_tracing.c:588\n#1  0x000055555594c141 in _PyEval_SetTrace (tstate=tstate@entry=0x555555e29760, func=func@entry=0x5555559984b0 <trace_trampoline>,\n    arg=arg@entry=()) at Python/legacy_tracing.c:610\n#2  0x000055555584cd32 in PyEval_SetTraceAllThreads (func=0x5555559984b0 <trace_trampoline>, arg=()) at Python/ceval.c:2473\n#3  0x000055555599698e in sys__settraceallthreads (module=<optimized out>, arg=<optimized out>) at ./Python/sysmodule.c:1187\n#4  0x00005555557110b6 in cfunction_vectorcall_O (\n    func=<built-in method _settraceallthreads of module object at remote 0x20000259930>, args=<optimized out>,\n    nargsf=<optimized out>, kwnames=<optimized out>) at Objects/methodobject.c:537\n#5  0x00005555556817dd in _PyObject_VectorcallTstate (tstate=0x555555d8a858 <_PyRuntime+361432>,\n    callable=<built-in method _settraceallthreads of module object at remote 0x20000259930>, args=0x7fffffffd708,\n    nargsf=9223372036854775809, kwnames=0x0) at ./Include/internal/pycore_call.h:169\n#6  0x00005555556818fc in PyObject_Vectorcall (\n    callable=callable@entry=<built-in method _settraceallthreads of module object at remote 0x20000259930>,\n    args=args@entry=0x7fffffffd708, nargsf=<optimized out>, kwnames=kwnames@entry=0x0) at Objects/call.c:327\n#7  0x000055555585575c in _PyEval_EvalFrameDefault (tstate=tstate@entry=0x555555d8a858 <_PyRuntime+361432>, frame=0x7ffff7fb0098,\n    frame@entry=0x7ffff7fb0020, throwflag=throwflag@entry=0) at Python/generated_cases.c.h:1434\n#8  0x0000555555888b6d in _PyEval_EvalFrame (throwflag=0, frame=0x7ffff7fb0020, tstate=0x555555d8a858 <_PyRuntime+361432>)\n    at ./Include/internal/pycore_ceval.h:119\n```\n\n### CPython versions tested on:\n\nCPython main branch\n\n### Operating systems tested on:\n\nLinux\n\n### Output from running 'python -VV' on the command line:\n\nPython 3.14.0a6+ experimental free-threading build (heads/main:0f04f2456a2, Apr  8 2025, 11:49:24) [GCC 11.4.0]\n\n<!-- gh-linked-prs -->\n### Linked PRs\n* gh-132298\n<!-- /gh-linked-prs -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132296/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132296/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132294", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132294/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132294/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132294/events", "html_url": "https://github.com/python/cpython/pull/132294", "id": 2981123353, "node_id": "PR_kwDOBN0Z8c6R29qB", "number": 132294, "title": "gh-127495: append to history file after every successful statement in new repl", "user": {"login": "skirpichev", "id": 2155800, "node_id": "MDQ6VXNlcjIxNTU4MDA=", "avatar_url": "https://avatars.githubusercontent.com/u/2155800?v=4", "gravatar_id": "", "url": "https://api.github.com/users/skirpichev", "html_url": "https://github.com/skirpichev", "followers_url": "https://api.github.com/users/skirpichev/followers", "following_url": "https://api.github.com/users/skirpichev/following{/other_user}", "gists_url": "https://api.github.com/users/skirpichev/gists{/gist_id}", "starred_url": "https://api.github.com/users/skirpichev/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/skirpichev/subscriptions", "organizations_url": "https://api.github.com/users/skirpichev/orgs", "repos_url": "https://api.github.com/users/skirpichev/repos", "events_url": "https://api.github.com/users/skirpichev/events{/privacy}", "received_events_url": "https://api.github.com/users/skirpichev/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 666979821, "node_id": "MDU6TGFiZWw2NjY5Nzk4MjE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20review", "name": "awaiting review", "color": "fbca04", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-08T23:09:26Z", "updated_at": "2025-04-09T01:22:07Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132294", "html_url": "https://github.com/python/cpython/pull/132294", "diff_url": "https://github.com/python/cpython/pull/132294.diff", "patch_url": "https://github.com/python/cpython/pull/132294.patch", "merged_at": null}, "body": "<!--\r\nThanks for your contribution!\r\nPlease read this comment in its entirety. It's quite important.\r\n\r\n# Pull Request title\r\n\r\nIt should be in the following format:\r\n\r\n```\r\ngh-NNNNN: Summary of the changes made\r\n```\r\n\r\nWhere: gh-NNNNN refers to the GitHub issue number.\r\n\r\nMost PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.\r\n\r\n# Backport Pull Request title\r\n\r\nIf this is a backport PR (PR made against branches other than `main`),\r\nplease ensure that the PR title is in the following format:\r\n\r\n```\r\n[X.Y] <title from the original PR> (GH-NNNN)\r\n```\r\n\r\nWhere: [X.Y] is the branch name, e.g. [3.6].\r\n\r\nGH-NNNN refers to the PR number from `main`.\r\n\r\n-->\r\n\n\n<!-- gh-issue-number: gh-127495 -->\n* Issue: gh-127495\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132294/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132294/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132292", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132292/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132292/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132292/events", "html_url": "https://github.com/python/cpython/issues/132292", "id": 2980818146, "node_id": "I_kwDOBN0Z8c6xq6zi", "number": 132292, "title": "Contradicting requirements in collections.abc.Sequence docs and typing information", "user": {"login": "502E532E", "id": 74902135, "node_id": "MDQ6VXNlcjc0OTAyMTM1", "avatar_url": "https://avatars.githubusercontent.com/u/74902135?v=4", "gravatar_id": "", "url": "https://api.github.com/users/502E532E", "html_url": "https://github.com/502E532E", "followers_url": "https://api.github.com/users/502E532E/followers", "following_url": "https://api.github.com/users/502E532E/following{/other_user}", "gists_url": "https://api.github.com/users/502E532E/gists{/gist_id}", "starred_url": "https://api.github.com/users/502E532E/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/502E532E/subscriptions", "organizations_url": "https://api.github.com/users/502E532E/orgs", "repos_url": "https://api.github.com/users/502E532E/repos", "events_url": "https://api.github.com/users/502E532E/events{/privacy}", "received_events_url": "https://api.github.com/users/502E532E/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-08T20:00:09Z", "updated_at": "2025-04-08T20:00:09Z", "closed_at": null, "author_association": "NONE", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "# Documentation\n\nThe [collections.abc.Sequence](https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence) docs state that `__getitem__` needs to be implemented, and link to some general [`__getitem__` docs](https://docs.python.org/3/reference/datamodel.html#object.__getitem__). It states\n\n> For [sequence](https://docs.python.org/3/glossary.html#term-sequence) types, the accepted keys should be integers. **Optionally**, they may support [slice](https://docs.python.org/3/library/functions.html#slice) objects as well.\n\nBut the corresponding [typing information](https://github.com/python/typeshed/blob/dbe6cb5cfb016de85143f9cb5ae772561e8982f6/stdlib/typing.pyi#L583) **requires** slice support.\n\nTo resolve this, either the docs or the typing information should be adapted.\n\n## Related\nhttps://github.com/python/cpython/issues/120802 (general `__getitem__` docs improvements)\nMy corresponding question and [answer](https://stackoverflow.com/a/79518256/16563251) on Stackoverflow\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132292/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132292/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132291", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132291/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132291/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132291/events", "html_url": "https://github.com/python/cpython/pull/132291", "id": 2980805348, "node_id": "PR_kwDOBN0Z8c6R142H", "number": 132291, "title": "Escape hyphens in manpage", "user": {"login": "stefanor", "id": 442117, "node_id": "MDQ6VXNlcjQ0MjExNw==", "avatar_url": "https://avatars.githubusercontent.com/u/442117?v=4", "gravatar_id": "", "url": "https://api.github.com/users/stefanor", "html_url": "https://github.com/stefanor", "followers_url": "https://api.github.com/users/stefanor/followers", "following_url": "https://api.github.com/users/stefanor/following{/other_user}", "gists_url": "https://api.github.com/users/stefanor/gists{/gist_id}", "starred_url": "https://api.github.com/users/stefanor/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/stefanor/subscriptions", "organizations_url": "https://api.github.com/users/stefanor/orgs", "repos_url": "https://api.github.com/users/stefanor/repos", "events_url": "https://api.github.com/users/stefanor/events{/privacy}", "received_events_url": "https://api.github.com/users/stefanor/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}, {"id": 647202665, "node_id": "MDU6TGFiZWw2NDcyMDI2NjU=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20issue", "name": "skip issue", "color": "000000", "default": false, "description": null}, {"id": 666979821, "node_id": "MDU6TGFiZWw2NjY5Nzk4MjE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20review", "name": "awaiting review", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 4, "created_at": "2025-04-08T19:53:17Z", "updated_at": "2025-04-09T12:40:48Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132291", "html_url": "https://github.com/python/cpython/pull/132291", "diff_url": "https://github.com/python/cpython/pull/132291.diff", "patch_url": "https://github.com/python/cpython/pull/132291.patch", "merged_at": null}, "body": "Based on the patch submitted in https://bugs.debian.org/1101406 as the result of running a linter on the manpage.", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132291/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132291/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132290", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132290/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132290/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132290/events", "html_url": "https://github.com/python/cpython/pull/132290", "id": 2980720311, "node_id": "PR_kwDOBN0Z8c6R1mil", "number": 132290, "title": "Add lock-free set contains implemention", "user": {"login": "nascheme", "id": 690853, "node_id": "MDQ6VXNlcjY5MDg1Mw==", "avatar_url": "https://avatars.githubusercontent.com/u/690853?v=4", "gravatar_id": "", "url": "https://api.github.com/users/nascheme", "html_url": "https://github.com/nascheme", "followers_url": "https://api.github.com/users/nascheme/followers", "following_url": "https://api.github.com/users/nascheme/following{/other_user}", "gists_url": "https://api.github.com/users/nascheme/gists{/gist_id}", "starred_url": "https://api.github.com/users/nascheme/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/nascheme/subscriptions", "organizations_url": "https://api.github.com/users/nascheme/orgs", "repos_url": "https://api.github.com/users/nascheme/repos", "events_url": "https://api.github.com/users/nascheme/events{/privacy}", "received_events_url": "https://api.github.com/users/nascheme/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-08T19:08:25Z", "updated_at": "2025-04-08T21:52:01Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": true, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132290", "html_url": "https://github.com/python/cpython/pull/132290", "diff_url": "https://github.com/python/cpython/pull/132290.diff", "patch_url": "https://github.com/python/cpython/pull/132290.patch", "merged_at": null}, "body": "This roughly follows what was done for dictobject to make a lock-free lookup operation.  On a benchmark running `set.__contains__` in a tight loop, this is 1.35x faster on my computer.  In the bm_deepcopy benchmark, the gains are modest, between 1 to 2% faster.\r\n\r\nSummary of changes:\r\n\r\n- refactor `set_lookkey()` into `set_do_lookup()` which now takes a function pointer that does the entry comparison.  This is similar to dictobject and `do_lookup()`.  In an optimized build, the comparison function  is inlined and there should be no performance cost to this.\r\n- change `set_do_lookup` to return a status separately from the entry value.  \r\n- define macros to load/store the various set and setentry members (e.g. LOAD_HASH, STORE_HASH).  These use the required atomic operations for a free-threaded build.\r\n- use a deferred free on the set table array, if shared (only on free-threaded build, normal build always does an immediate free)\r\n- for free-threaded build, use explicit for loop to zero the table, rather than `memcpy()`.  \r\n- when mutating the set, assign `so->table` to NULL while the change is a happening.   Assign the real table array after the change is done.\r\n\r\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132290/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132290/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132289", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132289/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132289/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132289/events", "html_url": "https://github.com/python/cpython/pull/132289", "id": 2980694182, "node_id": "PR_kwDOBN0Z8c6R1g5y", "number": 132289, "title": "GH-131798: Remove JIT guards for `dict`, `frozenset`, `list`, `set`, and `tuple`", "user": {"login": "brandtbucher", "id": 40968415, "node_id": "MDQ6VXNlcjQwOTY4NDE1", "avatar_url": "https://avatars.githubusercontent.com/u/40968415?v=4", "gravatar_id": "", "url": "https://api.github.com/users/brandtbucher", "html_url": "https://github.com/brandtbucher", "followers_url": "https://api.github.com/users/brandtbucher/followers", "following_url": "https://api.github.com/users/brandtbucher/following{/other_user}", "gists_url": "https://api.github.com/users/brandtbucher/gists{/gist_id}", "starred_url": "https://api.github.com/users/brandtbucher/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/brandtbucher/subscriptions", "organizations_url": "https://api.github.com/users/brandtbucher/orgs", "repos_url": "https://api.github.com/users/brandtbucher/repos", "events_url": "https://api.github.com/users/brandtbucher/events{/privacy}", "received_events_url": "https://api.github.com/users/brandtbucher/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 537029789, "node_id": "MDU6TGFiZWw1MzcwMjk3ODk=", "url": "https://api.github.com/repos/python/cpython/labels/performance", "name": "performance", "color": "0052cc", "default": false, "description": "Performance or resource usage"}, {"id": 666980059, "node_id": "MDU6TGFiZWw2NjY5ODAwNTk=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20merge", "name": "awaiting merge", "color": "fbca04", "default": false, "description": null}, {"id": 4018668718, "node_id": "LA_kwDOBN0Z8c7viASu", "url": "https://api.github.com/repos/python/cpython/labels/interpreter-core", "name": "interpreter-core", "color": "09fc59", "default": false, "description": "(Objects, Python, Grammar, and Parser dirs)"}, {"id": 7018075156, "node_id": "LA_kwDOBN0Z8c8AAAABok9UFA", "url": "https://api.github.com/repos/python/cpython/labels/topic-JIT", "name": "topic-JIT", "color": "0052cc", "default": false, "description": ""}], "state": "open", "locked": false, "assignee": {"login": "brandtbucher", "id": 40968415, "node_id": "MDQ6VXNlcjQwOTY4NDE1", "avatar_url": "https://avatars.githubusercontent.com/u/40968415?v=4", "gravatar_id": "", "url": "https://api.github.com/users/brandtbucher", "html_url": "https://github.com/brandtbucher", "followers_url": "https://api.github.com/users/brandtbucher/followers", "following_url": "https://api.github.com/users/brandtbucher/following{/other_user}", "gists_url": "https://api.github.com/users/brandtbucher/gists{/gist_id}", "starred_url": "https://api.github.com/users/brandtbucher/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/brandtbucher/subscriptions", "organizations_url": "https://api.github.com/users/brandtbucher/orgs", "repos_url": "https://api.github.com/users/brandtbucher/repos", "events_url": "https://api.github.com/users/brandtbucher/events{/privacy}", "received_events_url": "https://api.github.com/users/brandtbucher/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "assignees": [{"login": "brandtbucher", "id": 40968415, "node_id": "MDQ6VXNlcjQwOTY4NDE1", "avatar_url": "https://avatars.githubusercontent.com/u/40968415?v=4", "gravatar_id": "", "url": "https://api.github.com/users/brandtbucher", "html_url": "https://github.com/brandtbucher", "followers_url": "https://api.github.com/users/brandtbucher/followers", "following_url": "https://api.github.com/users/brandtbucher/following{/other_user}", "gists_url": "https://api.github.com/users/brandtbucher/gists{/gist_id}", "starred_url": "https://api.github.com/users/brandtbucher/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/brandtbucher/subscriptions", "organizations_url": "https://api.github.com/users/brandtbucher/orgs", "repos_url": "https://api.github.com/users/brandtbucher/repos", "events_url": "https://api.github.com/users/brandtbucher/events{/privacy}", "received_events_url": "https://api.github.com/users/brandtbucher/received_events", "type": "User", "user_view_type": "public", "site_admin": false}], "milestone": null, "comments": 2, "created_at": "2025-04-08T18:55:33Z", "updated_at": "2025-04-08T20:02:30Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132289", "html_url": "https://github.com/python/cpython/pull/132289", "diff_url": "https://github.com/python/cpython/pull/132289.diff", "patch_url": "https://github.com/python/cpython/pull/132289.patch", "merged_at": null}, "body": "According to the [stats](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20250401-3.14.0a6%2B-99d6288-JIT/bm-20250401-azure-x86_64-brandtbucher-jit_guard_dict_tuple-3.14.0a6%2B-99d6288-pystats-vs-base.md), this removes:\r\n- 144 million `tuple` guards (25%)\r\n- 280 million `set`/`frozenset` guards (19%)\r\n- 484 million `dict` guards (42%)\r\n- 1.3 billion `list` guards (34%)\r\n\r\nIt also fixes a bug that I encountered in the optimizer's logic for `_UNPACK_SEQUENCE_TUPLE`, which puts items on the stack in *reverse* order.\n\n<!-- gh-issue-number: gh-131798 -->\n* Issue: gh-131798\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132289/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132289/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132285", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132285/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132285/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132285/events", "html_url": "https://github.com/python/cpython/issues/132285", "id": 2980609558, "node_id": "I_kwDOBN0Z8c6xqH4W", "number": 132285, "title": "`__annotate__` corner case when deleting `__annotations__`", "user": {"login": "sobolevn", "id": 4660275, "node_id": "MDQ6VXNlcjQ2NjAyNzU=", "avatar_url": "https://avatars.githubusercontent.com/u/4660275?v=4", "gravatar_id": "", "url": "https://api.github.com/users/sobolevn", "html_url": "https://github.com/sobolevn", "followers_url": "https://api.github.com/users/sobolevn/followers", "following_url": "https://api.github.com/users/sobolevn/following{/other_user}", "gists_url": "https://api.github.com/users/sobolevn/gists{/gist_id}", "starred_url": "https://api.github.com/users/sobolevn/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/sobolevn/subscriptions", "organizations_url": "https://api.github.com/users/sobolevn/orgs", "repos_url": "https://api.github.com/users/sobolevn/repos", "events_url": "https://api.github.com/users/sobolevn/events{/privacy}", "received_events_url": "https://api.github.com/users/sobolevn/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 537044141, "node_id": "MDU6TGFiZWw1MzcwNDQxNDE=", "url": "https://api.github.com/repos/python/cpython/labels/tests", "name": "tests", "color": "09fc59", "default": false, "description": "Tests in the Lib/test dir"}, {"id": 4030784939, "node_id": "LA_kwDOBN0Z8c7wQOWr", "url": "https://api.github.com/repos/python/cpython/labels/topic-typing", "name": "topic-typing", "color": "0052cc", "default": false, "description": ""}], "state": "open", "locked": false, "assignee": {"login": "sobolevn", "id": 4660275, "node_id": "MDQ6VXNlcjQ2NjAyNzU=", "avatar_url": "https://avatars.githubusercontent.com/u/4660275?v=4", "gravatar_id": "", "url": "https://api.github.com/users/sobolevn", "html_url": "https://github.com/sobolevn", "followers_url": "https://api.github.com/users/sobolevn/followers", "following_url": "https://api.github.com/users/sobolevn/following{/other_user}", "gists_url": "https://api.github.com/users/sobolevn/gists{/gist_id}", "starred_url": "https://api.github.com/users/sobolevn/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/sobolevn/subscriptions", "organizations_url": "https://api.github.com/users/sobolevn/orgs", "repos_url": "https://api.github.com/users/sobolevn/repos", "events_url": "https://api.github.com/users/sobolevn/events{/privacy}", "received_events_url": "https://api.github.com/users/sobolevn/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "assignees": [{"login": "sobolevn", "id": 4660275, "node_id": "MDQ6VXNlcjQ2NjAyNzU=", "avatar_url": "https://avatars.githubusercontent.com/u/4660275?v=4", "gravatar_id": "", "url": "https://api.github.com/users/sobolevn", "html_url": "https://github.com/sobolevn", "followers_url": "https://api.github.com/users/sobolevn/followers", "following_url": "https://api.github.com/users/sobolevn/following{/other_user}", "gists_url": "https://api.github.com/users/sobolevn/gists{/gist_id}", "starred_url": "https://api.github.com/users/sobolevn/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/sobolevn/subscriptions", "organizations_url": "https://api.github.com/users/sobolevn/orgs", "repos_url": "https://api.github.com/users/sobolevn/repos", "events_url": "https://api.github.com/users/sobolevn/events{/privacy}", "received_events_url": "https://api.github.com/users/sobolevn/received_events", "type": "User", "user_view_type": "public", "site_admin": false}], "milestone": null, "comments": 0, "created_at": "2025-04-08T18:16:27Z", "updated_at": "2025-04-08T18:24:22Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "Consider this example:\n\n```python\n>>> class A:\n...      def __annotate__(format):\n...           print(f'{format=}')\n...           return {'a': int}\n...           \n>>> A.__annotations__\nformat=1\n{'a': <class 'int'>}\n>>> A.__annotations__\n{'a': <class 'int'>}\n>>> del A.__annotations__\n>>> A.__annotations__\nformat=1\n{'a': <class 'int'>}\n```\n\nIn this particular case `__annotate__` is called twice. It works correctly in my opinion. But, I can't find any tests for such cases. Should I add tests for this?\n\nCC @JelleZijlstra \n\n<!-- gh-linked-prs -->\n### Linked PRs\n* gh-132286\n<!-- /gh-linked-prs -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132285/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132285/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132284", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132284/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132284/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132284/events", "html_url": "https://github.com/python/cpython/issues/132284", "id": 2980607311, "node_id": "I_kwDOBN0Z8c6xqHVP", "number": 132284, "title": "Py_TYPE(lhs)->tp_as_mapping->mp_subscript is not the same as Dict_Type.tp_as_mapping->mp_subscript when it should be", "user": {"login": "iritkatriel", "id": 1055913, "node_id": "MDQ6VXNlcjEwNTU5MTM=", "avatar_url": "https://avatars.githubusercontent.com/u/1055913?v=4", "gravatar_id": "", "url": "https://api.github.com/users/iritkatriel", "html_url": "https://github.com/iritkatriel", "followers_url": "https://api.github.com/users/iritkatriel/followers", "following_url": "https://api.github.com/users/iritkatriel/following{/other_user}", "gists_url": "https://api.github.com/users/iritkatriel/gists{/gist_id}", "starred_url": "https://api.github.com/users/iritkatriel/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/iritkatriel/subscriptions", "organizations_url": "https://api.github.com/users/iritkatriel/orgs", "repos_url": "https://api.github.com/users/iritkatriel/repos", "events_url": "https://api.github.com/users/iritkatriel/events{/privacy}", "received_events_url": "https://api.github.com/users/iritkatriel/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 4018668718, "node_id": "LA_kwDOBN0Z8c7viASu", "url": "https://api.github.com/repos/python/cpython/labels/interpreter-core", "name": "interpreter-core", "color": "09fc59", "default": false, "description": "(Objects, Python, Grammar, and Parser dirs)"}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 1, "created_at": "2025-04-08T18:15:25Z", "updated_at": "2025-04-08T18:25:41Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "body": "\nA dict subclass which does not override `__getitem__` (like `collections.Counter`) should have `Dict_Type.tp_as_mapping->mp_subscript` as its `Py_TYPE(lhs)->tp_as_mapping->mp_subscript`, but they are not the same.\n\nCC @ericsnowcurrently .", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132284/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132284/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132283", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132283/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132283/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132283/events", "html_url": "https://github.com/python/cpython/pull/132283", "id": 2980530116, "node_id": "PR_kwDOBN0Z8c6R09KR", "number": 132283, "title": "gh-129463: Remove two attributes from ForwardRef equality", "user": {"login": "JelleZijlstra", "id": 906600, "node_id": "MDQ6VXNlcjkwNjYwMA==", "avatar_url": "https://avatars.githubusercontent.com/u/906600?v=4", "gravatar_id": "", "url": "https://api.github.com/users/JelleZijlstra", "html_url": "https://github.com/JelleZijlstra", "followers_url": "https://api.github.com/users/JelleZijlstra/followers", "following_url": "https://api.github.com/users/JelleZijlstra/following{/other_user}", "gists_url": "https://api.github.com/users/JelleZijlstra/gists{/gist_id}", "starred_url": "https://api.github.com/users/JelleZijlstra/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/JelleZijlstra/subscriptions", "organizations_url": "https://api.github.com/users/JelleZijlstra/orgs", "repos_url": "https://api.github.com/users/JelleZijlstra/repos", "events_url": "https://api.github.com/users/JelleZijlstra/events{/privacy}", "received_events_url": "https://api.github.com/users/JelleZijlstra/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 666979871, "node_id": "MDU6TGFiZWw2NjY5Nzk4NzE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20core%20review", "name": "awaiting core review", "color": "fbca04", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 3, "created_at": "2025-04-08T17:48:20Z", "updated_at": "2025-04-09T13:03:56Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132283", "html_url": "https://github.com/python/cpython/pull/132283", "diff_url": "https://github.com/python/cpython/pull/132283.diff", "patch_url": "https://github.com/python/cpython/pull/132283.patch", "merged_at": null}, "body": "* `__code__` is created on demand in the `__forward_code__` property\r\n* `__ast_node__` is used together with `__arg__` to compute `__forward_arg__`, which is already used by itself in the comparison\n\n<!-- gh-issue-number: gh-129463 -->\n* Issue: gh-129463\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132283/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132283/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132282", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132282/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132282/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132282/events", "html_url": "https://github.com/python/cpython/pull/132282", "id": 2980519870, "node_id": "PR_kwDOBN0Z8c6R06-e", "number": 132282, "title": "[3.13] gh-132246: Add PEP 688 to C Buffer Protocol docs (GH-132249)", "user": {"login": "miss-islington", "id": 31488909, "node_id": "MDQ6VXNlcjMxNDg4OTA5", "avatar_url": "https://avatars.githubusercontent.com/u/31488909?v=4", "gravatar_id": "", "url": "https://api.github.com/users/miss-islington", "html_url": "https://github.com/miss-islington", "followers_url": "https://api.github.com/users/miss-islington/followers", "following_url": "https://api.github.com/users/miss-islington/following{/other_user}", "gists_url": "https://api.github.com/users/miss-islington/gists{/gist_id}", "starred_url": "https://api.github.com/users/miss-islington/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/miss-islington/subscriptions", "organizations_url": "https://api.github.com/users/miss-islington/orgs", "repos_url": "https://api.github.com/users/miss-islington/repos", "events_url": "https://api.github.com/users/miss-islington/events{/privacy}", "received_events_url": "https://api.github.com/users/miss-islington/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 538516109, "node_id": "MDU6TGFiZWw1Mzg1MTYxMDk=", "url": "https://api.github.com/repos/python/cpython/labels/docs", "name": "docs", "color": "09fc59", "default": false, "description": "Documentation in the Doc dir"}, {"id": 666979821, "node_id": "MDU6TGFiZWw2NjY5Nzk4MjE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20review", "name": "awaiting review", "color": "fbca04", "default": false, "description": null}, {"id": 667924166, "node_id": "MDU6TGFiZWw2Njc5MjQxNjY=", "url": "https://api.github.com/repos/python/cpython/labels/skip%20news", "name": "skip news", "color": "000000", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 0, "created_at": "2025-04-08T17:43:37Z", "updated_at": "2025-04-08T17:43:50Z", "closed_at": null, "author_association": "CONTRIBUTOR", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132282", "html_url": "https://github.com/python/cpython/pull/132282", "diff_url": "https://github.com/python/cpython/pull/132282.diff", "patch_url": "https://github.com/python/cpython/pull/132282.patch", "merged_at": null}, "body": "(cherry picked from commit 8421b648e91981e393a740dd9fb7b7dbf4cf07dc)\n\nCo-authored-by: Cody Maloney <cmaloney@users.noreply.github.com>\n\n<!-- gh-issue-number: gh-132246 -->\n* Issue: gh-132246\n<!-- /gh-issue-number -->\n\r\n\r\n<!-- readthedocs-preview cpython-previews start -->\r\n----\n\ud83d\udcda Documentation preview \ud83d\udcda: https://cpython-previews--132282.org.readthedocs.build/\n\r\n<!-- readthedocs-preview cpython-previews end -->", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132282/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132282/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132280", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132280/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132280/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132280/events", "html_url": "https://github.com/python/cpython/pull/132280", "id": 2980351135, "node_id": "PR_kwDOBN0Z8c6R0Vul", "number": 132280, "title": "GH-124715: Move trashcan mechanism into `Py_Dealloc`", "user": {"login": "markshannon", "id": 9448417, "node_id": "MDQ6VXNlcjk0NDg0MTc=", "avatar_url": "https://avatars.githubusercontent.com/u/9448417?v=4", "gravatar_id": "", "url": "https://api.github.com/users/markshannon", "html_url": "https://github.com/markshannon", "followers_url": "https://api.github.com/users/markshannon/followers", "following_url": "https://api.github.com/users/markshannon/following{/other_user}", "gists_url": "https://api.github.com/users/markshannon/gists{/gist_id}", "starred_url": "https://api.github.com/users/markshannon/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/markshannon/subscriptions", "organizations_url": "https://api.github.com/users/markshannon/orgs", "repos_url": "https://api.github.com/users/markshannon/repos", "events_url": "https://api.github.com/users/markshannon/events{/privacy}", "received_events_url": "https://api.github.com/users/markshannon/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 666979871, "node_id": "MDU6TGFiZWw2NjY5Nzk4NzE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20core%20review", "name": "awaiting core review", "color": "fbca04", "default": false, "description": null}], "state": "open", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 3, "created_at": "2025-04-08T16:36:58Z", "updated_at": "2025-04-09T15:00:06Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132280", "html_url": "https://github.com/python/cpython/pull/132280", "diff_url": "https://github.com/python/cpython/pull/132280.diff", "patch_url": "https://github.com/python/cpython/pull/132280.patch", "merged_at": null}, "body": "<!--\r\nThanks for your contribution!\r\nPlease read this comment in its entirety. It's quite important.\r\n\r\n# Pull Request title\r\n\r\nIt should be in the following format:\r\n\r\n```\r\ngh-NNNNN: Summary of the changes made\r\n```\r\n\r\nWhere: gh-NNNNN refers to the GitHub issue number.\r\n\r\nMost PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.\r\n\r\n# Backport Pull Request title\r\n\r\nIf this is a backport PR (PR made against branches other than `main`),\r\nplease ensure that the PR title is in the following format:\r\n\r\n```\r\n[X.Y] <title from the original PR> (GH-NNNN)\r\n```\r\n\r\nWhere: [X.Y] is the branch name, e.g. [3.6].\r\n\r\nGH-NNNN refers to the PR number from `main`.\r\n\r\n-->\r\n\n\n<!-- gh-issue-number: gh-124715 -->\n* Issue: gh-124715\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132280/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132280/timeline", "performed_via_github_app": null, "state_reason": null}, {"url": "https://api.github.com/repos/python/cpython/issues/132278", "repository_url": "https://api.github.com/repos/python/cpython", "labels_url": "https://api.github.com/repos/python/cpython/issues/132278/labels{/name}", "comments_url": "https://api.github.com/repos/python/cpython/issues/132278/comments", "events_url": "https://api.github.com/repos/python/cpython/issues/132278/events", "html_url": "https://github.com/python/cpython/pull/132278", "id": 2980277090, "node_id": "PR_kwDOBN0Z8c6R0FiD", "number": 132278, "title": "GH-131798: Skip `self`/`NULL` checks for some known non-methods", "user": {"login": "brandtbucher", "id": 40968415, "node_id": "MDQ6VXNlcjQwOTY4NDE1", "avatar_url": "https://avatars.githubusercontent.com/u/40968415?v=4", "gravatar_id": "", "url": "https://api.github.com/users/brandtbucher", "html_url": "https://github.com/brandtbucher", "followers_url": "https://api.github.com/users/brandtbucher/followers", "following_url": "https://api.github.com/users/brandtbucher/following{/other_user}", "gists_url": "https://api.github.com/users/brandtbucher/gists{/gist_id}", "starred_url": "https://api.github.com/users/brandtbucher/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/brandtbucher/subscriptions", "organizations_url": "https://api.github.com/users/brandtbucher/orgs", "repos_url": "https://api.github.com/users/brandtbucher/repos", "events_url": "https://api.github.com/users/brandtbucher/events{/privacy}", "received_events_url": "https://api.github.com/users/brandtbucher/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "labels": [{"id": 537029789, "node_id": "MDU6TGFiZWw1MzcwMjk3ODk=", "url": "https://api.github.com/repos/python/cpython/labels/performance", "name": "performance", "color": "0052cc", "default": false, "description": "Performance or resource usage"}, {"id": 666979871, "node_id": "MDU6TGFiZWw2NjY5Nzk4NzE=", "url": "https://api.github.com/repos/python/cpython/labels/awaiting%20core%20review", "name": "awaiting core review", "color": "fbca04", "default": false, "description": null}, {"id": 4018668718, "node_id": "LA_kwDOBN0Z8c7viASu", "url": "https://api.github.com/repos/python/cpython/labels/interpreter-core", "name": "interpreter-core", "color": "09fc59", "default": false, "description": "(Objects, Python, Grammar, and Parser dirs)"}], "state": "open", "locked": false, "assignee": {"login": "brandtbucher", "id": 40968415, "node_id": "MDQ6VXNlcjQwOTY4NDE1", "avatar_url": "https://avatars.githubusercontent.com/u/40968415?v=4", "gravatar_id": "", "url": "https://api.github.com/users/brandtbucher", "html_url": "https://github.com/brandtbucher", "followers_url": "https://api.github.com/users/brandtbucher/followers", "following_url": "https://api.github.com/users/brandtbucher/following{/other_user}", "gists_url": "https://api.github.com/users/brandtbucher/gists{/gist_id}", "starred_url": "https://api.github.com/users/brandtbucher/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/brandtbucher/subscriptions", "organizations_url": "https://api.github.com/users/brandtbucher/orgs", "repos_url": "https://api.github.com/users/brandtbucher/repos", "events_url": "https://api.github.com/users/brandtbucher/events{/privacy}", "received_events_url": "https://api.github.com/users/brandtbucher/received_events", "type": "User", "user_view_type": "public", "site_admin": false}, "assignees": [{"login": "brandtbucher", "id": 40968415, "node_id": "MDQ6VXNlcjQwOTY4NDE1", "avatar_url": "https://avatars.githubusercontent.com/u/40968415?v=4", "gravatar_id": "", "url": "https://api.github.com/users/brandtbucher", "html_url": "https://github.com/brandtbucher", "followers_url": "https://api.github.com/users/brandtbucher/followers", "following_url": "https://api.github.com/users/brandtbucher/following{/other_user}", "gists_url": "https://api.github.com/users/brandtbucher/gists{/gist_id}", "starred_url": "https://api.github.com/users/brandtbucher/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/brandtbucher/subscriptions", "organizations_url": "https://api.github.com/users/brandtbucher/orgs", "repos_url": "https://api.github.com/users/brandtbucher/repos", "events_url": "https://api.github.com/users/brandtbucher/events{/privacy}", "received_events_url": "https://api.github.com/users/brandtbucher/received_events", "type": "User", "user_view_type": "public", "site_admin": false}], "milestone": null, "comments": 5, "created_at": "2025-04-08T16:08:07Z", "updated_at": "2025-04-09T08:52:55Z", "closed_at": null, "author_association": "MEMBER", "type": null, "sub_issues_summary": {"total": 0, "completed": 0, "percent_completed": 0}, "active_lock_reason": null, "draft": false, "pull_request": {"url": "https://api.github.com/repos/python/cpython/pulls/132278", "html_url": "https://github.com/python/cpython/pull/132278", "diff_url": "https://github.com/python/cpython/pull/132278.diff", "patch_url": "https://github.com/python/cpython/pull/132278.patch", "merged_at": null}, "body": "`CALL_TYPE_1`, `CALL_STR_1`, `CALL_TUPLE_1`, `CALL_LEN`, and `CALL_ISINSTANCE` will never call a method descriptor, which means that we can just assert that `self_or_null` is `NULL` instead of checking and \"adjusting\" the arguments each time.\r\n\r\n`CALL_BUILTIN_CLASS` should never call a method descriptor, but I'm not sure that we can gurantee it. So I've just made this a `DEOPT_IF` instead.\n\n<!-- gh-issue-number: gh-131798 -->\n* Issue: gh-131798\n<!-- /gh-issue-number -->\n", "closed_by": null, "reactions": {"url": "https://api.github.com/repos/python/cpython/issues/132278/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/python/cpython/issues/132278/timeline", "performed_via_github_app": null, "state_reason": null}]